src/main/java/com/cicerone/util/io/HTMLGenerator.java

```diff
@@ -0,0 +1,166 @@
+
+            String title = properties[0];
+            String slug = properties[1].trim();
+            int order = properties[2].isEmpty() ? 0 : Integer.parseInt(properties[2]);
+            String description = properties[3];
+            boolean disabled = !properties[4].equals("ATIVA");
+            String iconPath = properties[5];
+            String colorHexCode = properties[6];
+
+            return new Category(title, slug, order, description, disabled, iconPath, colorHexCode);
+        });
+
+        Map<String, Category> categoriesMap = categories.stream().collect(Collectors.toMap(Category::getSlug, Function.identity()));
+
+        this.subcategories = CSVDataLoader.load(this.sourceDirPath.concat("/subcategories.csv"), (properties) -> {
+
+            String title = properties[0];
+            String slug = properties[1].trim();
+            int order = properties[2].isEmpty() ? 0 : Integer.parseInt(properties[2]);
+            String description = properties[3];
+            boolean disabled = !properties[4].equals("ATIVA");
+            Category parentCategory = categoriesMap.get(properties[5]);
+
+            return new Category(title, slug, order, description, disabled, parentCategory);
+        });
+
+        Map<String, Category> subcategoriesMap = subcategories.stream().collect(Collectors.toMap(Category::getSlug, Function.identity()));
+
+        this.courses = CSVDataLoader.load(this.sourceDirPath.concat("/courses.csv"), (properties) -> {
+
+            String title = properties[0];
+            String slug = properties[1].trim();
+            int timeToFinishInHours = Integer.parseInt(properties[2]);
+            boolean disabled = !properties[3].equals("PÚBLICA");
+            String targetAudience = properties[4];
+            String instructor = properties[5];
+            String program = properties[6];
+            String skills = properties[7];
+            Category category = subcategoriesMap.get(properties[8]);
+
+            return new Course(title, slug, timeToFinishInHours, disabled, targetAudience, instructor, program, skills, category);
+        });
+
+    }
+
+    private Map<Category, List<Course>> mapCoursesToSubcategories(Category category) {
+
+        Map<Category, List<Course>> coursesBySub = subcategories.stream()
+                .filter(sub -> sub.getParentCategorySlug().equals(category.getSlug()))
+                .collect(Collectors.toMap(
+                        Function.identity(),
+                        sub -> courses.stream().filter(course -> course.getCategorySlug().equals(sub.getSlug())).collect(Collectors.toList())
+                ));
+
+        category.setCourseAmount(coursesBySub.values().stream().mapToInt(List::size).sum());
+
+        category.setTimeToFinishInHours(coursesBySub.values().stream().mapToInt(subCourses ->
+                subCourses.stream().mapToInt(Course::getTimeToFinishInHours).sum()).sum());
```

Comment:
Acho que isso ainda poderia estar encapsulado em `Category`.

Para pegar o total de cursos de uma categoria, faz dentro de `Category`.
E lá dentro, soma o total de cursos da subcategoria, que é implementado em `Subcategory`.
E dentro de `Subcategory`, pega o tamanho da lista de cursos.

Para o tempo de finalização total, faz algo semelhante. Categoria faz a soma do resultado da Subcategoria que, por sua vez, faz soma do que pegar de Curso.

Reply:
Beleza. Nesse caso, as categorias terão que manter referência para suas subcategorias e estas terão que manter referência para seus cursos, né? Vou implementar dessa forma :)

Reply:
Dá pra deixar como está e criar um outro objeto com esses dados. Talvez um objeto que represente o relatório. Aí poderia ter um `Map<Category, List<Category>>` e `Map<Category, List<Course>>`. E os totais não ficariam na `Category`.

Reply:
Acabei achando melhor refazer a modelagem de categorias e subcategorias, pois elas de fato desempenham papéis diferentes.